File: src/main/java/toby/command/commands/music/QueueCommand.java
package toby.command.commands.music;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.requests.restaction.WebhookMessageCreateAction;
import toby.command.CommandContext;
import toby.jpa.dto.UserDto;
import toby.lavaplayer.PlayerManager;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;

import static toby.command.ICommand.invokeDeleteOnMessageResponse;
import static toby.helpers.MusicPlayerHelper.formatTime;

public class QueueCommand implements IMusicCommand {

    @Override
    public void handle(CommandContext ctx, UserDto requestingUserDto, Integer deleteDelay) {
        handleMusicCommand(ctx, PlayerManager.getInstance(), requestingUserDto, deleteDelay);
    }

    @Override
    public void handleMusicCommand(CommandContext ctx, PlayerManager instance, UserDto requestingUserDto, Integer deleteDelay) {
        final SlashCommandInteractionEvent event = ctx.getEvent();
        event.deferReply().queue();
        final BlockingQueue<AudioTrack> queue = instance.getMusicManager(ctx.getGuild()).getScheduler().getQueue();

        if (!requestingUserDto.hasMusicPermission()) {
            sendErrorMessage(event, deleteDelay);
            return;
        }

        if (queue.isEmpty()) {
            event.getHook().sendMessage("The queue is currently empty").setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay));
            return;
        }

        final int trackCount = Math.min(queue.size(), 20);
        final List<AudioTrack> trackList = new ArrayList<>(queue);
        WebhookMessageCreateAction<Message> messageAction = event.getHook().sendMessage("**Current Queue:**\n");

        for (int i = 0; i < trackCount; i++) {
            final AudioTrack track = trackList.get(i);
            final AudioTrackInfo info = track.getInfo();

            messageAction.addContent("#")
                    .addContent(String.valueOf(i + 1))
                    .addContent(" `")
                    .addContent(String.valueOf(info.title))
                    .addContent(" by ")
                    .addContent(info.author)
                    .addContent("` [`")
                    .addContent(formatTime(track.getDuration()))
                    .addContent("`]\n");
        }

        if (trackList.size() > trackCount) {
            messageAction.addContent("And `")
                    .addContent(String.valueOf(trackList.size() - trackCount))
                    .addContent("` more...");
        }

        messageAction.setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay));
    }


    @Override
    public String getName() {
        return "queue";
    }

    @Override
    public String getDescription() {
        return "shows the queued up songs";
    }
}

File: src/main/kotlin/toby/command/commands/music/player/QueueCommand.kt
package toby.command.commands.music.player

import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import toby.command.CommandContext
import toby.command.ICommand.Companion.invokeDeleteOnMessageResponse
import toby.command.commands.music.IMusicCommand
import toby.helpers.MusicPlayerHelper
import toby.jpa.dto.UserDto
import toby.lavaplayer.PlayerManager
import kotlin.math.min

class QueueCommand : IMusicCommand {
    override fun handle(ctx: CommandContext, requestingUserDto: UserDto, deleteDelay: Int?) {
        handleMusicCommand(ctx, PlayerManager.instance, requestingUserDto, deleteDelay)
    }

    override fun handleMusicCommand(ctx: CommandContext, instance: PlayerManager, requestingUserDto: UserDto, deleteDelay: Int?) {
        val event = ctx.event
        event.deferReply(true).queue()
        val queue = instance.getMusicManager(ctx.guild).scheduler.queue
        if (!requestingUserDto.musicPermission) {
            sendErrorMessage(event, deleteDelay!!)
            return
        }
        if (queue.isEmpty()) {
            event.hook.sendMessage("The queue is currently empty").setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay!!))
            return
        }
        val trackCount = min(queue.size.toDouble(), 20.0).toInt()
        val trackList: List<AudioTrack?> = queue.toList()
        val messageAction = event.hook.sendMessage("**Current Queue:**\n")
        for (i in 0 until trackCount) {
            val track = trackList[i]
            val info = track?.info
            messageAction.addContent("#")
                    .addContent((i + 1).toString())
                    .addContent(" `")
                    .addContent(info?.title.toString())
                    .addContent(" by ")
                    .addContent(info?.author!!)
                    .addContent("` [`")
                    .addContent(MusicPlayerHelper.formatTime(track.duration))
                    .addContent("`]\n")
        }
        if (trackList.size > trackCount) {
            messageAction.addContent("And `")
                    .addContent((trackList.size - trackCount).toString())
                    .addContent("` more...")
        }
        messageAction.setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay!!))
    }

    override val name: String
        get() = "queue"
    override val description: String
        get() = "shows the queued up songs"
}
